### PR TITLE
Add `adjust_message_entities_to_utf_16` utility

### PR DESCRIPTION
**Pull request by Antares0982**
add  `adjust_message_entities_to_utf_16` for #4319

### CODE REVIEW

**Comment by Antares0982** [comment]
Is there anything else should be add/changed?

File: telegram/_messageentity.py [hunk 1]
```diff
@@ -140,6 +140,92 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+        when the text contains characters that are not in the Basic Multilingual Plane (BMP).
+
+        Examples:
+        Below is a snippet of code that demonstrates how to use this function to convert
+        entities from Unicode to UTF-16 space. The `unicode_entities` are calculated in
+        Unicode and the `utf_16_entities` are calculated in UTF-16.
+
+        ```python
+        text = "𠌕 bold 𝄢 italic underlined: 𝛙𝌢𑁍"
+        #         ^^^^   ^^^^^^             ^^^
```
Comment (Bibo-Joshi) [inline telegram/_messageentity.py:162]: it looks like the marking does not align. please make sure that this renders correctly in the html docs

File: telegram/_messageentity.py [hunk 2]
```diff
@@ -140,6 +140,92 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+            offsets_set.add(entity.offset + entity.length)
+        # get sorted offsets
+        offsets_list = sorted(offsets_set)
+        # indexes_map stores the index of each `cut_offset` in `offsets_list`
+        indexes_map = {cut_offset: i for i, cut_offset in enumerate(offsets_list)}
+        # calculate the length of each slice in utf-16
+        accumulated_length = 0
+        utf16_lengths = []
+        # what utf16_lengths stores:
+        # utf16_lengths[indexes_map[offset]] = len(text[:offset].encode("utf-16-le")) // 2
```
Comment (Bibo-Joshi) [inline telegram/_messageentity.py:203]: this comment is not particularly helpful. A code comment should explain what the meaning of the code is or why it is done. In this case, the meaning of the code would be something like `for each split_mark in unicode-space it stores the corresponding split_mark in utf_16-space`

File: telegram/_messageentity.py [hunk 3]
```diff
@@ -140,6 +140,92 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+        """
+        offsets_set = set()
+        # collect all offsets occured
+        for entity in entities:
+            offsets_set.add(entity.offset)
+            offsets_set.add(entity.offset + entity.length)
+        # get sorted offsets
+        offsets_list = sorted(offsets_set)
+        # indexes_map stores the index of each `cut_offset` in `offsets_list`
+        indexes_map = {cut_offset: i for i, cut_offset in enumerate(offsets_list)}
```
Comment (Bibo-Joshi) [inline telegram/_messageentity.py:198]: I think this map is superflous - all you need is a mapping from unidoce-positions to utf_16-positions. So something like 

```python
position_mapping: Dict[int, int] = {}
for split_mark in split_marks:
    # do computation
    position_mapping[split_mark] = accumulated_length
```

should already do the trick

File: telegram/_messageentity.py [hunk 4]
```diff
@@ -140,6 +140,92 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+            new_entity = MessageEntity(
+                type=entity.type,
+                offset=translated_offset,
+                length=translated_length,
+                url=entity.url,
+                user=entity.user,
+                language=entity.language,
+                custom_emoji_id=entity.custom_emoji_id,
+                api_kwargs={**entity.api_kwargs},
+            )
```
Comment (Bibo-Joshi) [inline telegram/_messageentity.py:225]: new_entity = copy.copy(entity)

```suggestion
            new_entity = copy.copy(entity)
            with new_entity._unfreeze():
                new_entity.offset = translated_offset
                new_entity.length = translated_length
```

`._unfreeze()` is an internal helper that allows us to override attributes after initialization if required.
Donig it this way prevents that we forget adding new arguments if they are added to `MessageEntity` :)

File: tests/test_messageentity.py [hunk 5]
```diff
@@ -96,3 +96,17 @@ def test_equality(self):
 
         assert a != d
         assert hash(a) != hash(d)
+
+    def test_fix_utf16(self):
```
Comment (Bibo-Joshi) [inline tests/test_messageentity.py:100]: please move to before `test_equality`

File: telegram/_messageentity.py [hunk 6]
```diff
@@ -140,6 +140,92 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+        # utf_16_entities[2]: offset=30, length=6
+        ```
+
+        Args:
+            text (:obj:`str`): The text that the entities belong to
+            entities (List[:class:`telegram.MessageEntity`]): Sequence of entities
+                with offset and length calculated in Unicode
+
+        Returns:
+            List[:class:`telegram.MessageEntity`]: Sequence of entities with offset and length
```
Comment (Bibo-Joshi) [inline telegram/_messageentity.py:187]:
```suggestion
            Sequence[:class:`telegram.MessageEntity`]: Sequence of entities with offset and length
```

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Hi, thank you very much for the PR! I left a number of comments. The overall logic is fine - thanks for catching the necessity for sorting! - but there are some things that can be improved :)

File: telegram/_messageentity.py [hunk 7]
```diff
@@ -140,6 +142,84 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+    @staticmethod
+    def adjust_message_entities_to_utf_16(
+        text: str, entities: Sequence["MessageEntity"]
+    ) -> Sequence["MessageEntity"]:
+        """Utility functionality for converting the offset and length of entities from
+        Unicode (`str`) to UTF-16 (`utf-16-le` encoded `bytes`).
+
+        Tip:
+            Only the offsets and lengths calulated in UTF-16 is acceptable by the Telegram Bot API.
+            If they are calculated using the Unicode string (`str` object), errors may occur
```
Comment (Bibo-Joshi) [inline telegram/_messageentity.py:154]:
```suggestion
            If they are calculated using the Unicode string (:obj:`str` object), errors may occur
```

File: telegram/_messageentity.py [hunk 8]
```diff
@@ -140,6 +142,84 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+            when the text contains characters that are not in the Basic Multilingual Plane (BMP).
+            For more information, see `Unicode <https://en.wikipedia.org/wiki/Unicode>` and
+            `Plane (Unicode) <https://en.wikipedia.org/wiki/Plane_(Unicode)>`.
+
+        Examples:
+            Below is a snippet of code that demonstrates how to use this function to convert
+            entities from Unicode to UTF-16 space. The `unicode_entities` are calculated in
+            Unicode and the `utf_16_entities` are calculated in UTF-16.
+
+        .. code-block:: python
```
Comment (Bibo-Joshi) [inline telegram/_messageentity.py:164]: can you indent this so that it's part of the "Example" admonition?

File: telegram/_messageentity.py [hunk 9]
```diff
@@ -140,6 +142,84 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+        Tip:
+            Only the offsets and lengths calulated in UTF-16 is acceptable by the Telegram Bot API.
+            If they are calculated using the Unicode string (`str` object), errors may occur
+            when the text contains characters that are not in the Basic Multilingual Plane (BMP).
+            For more information, see `Unicode <https://en.wikipedia.org/wiki/Unicode>` and
+            `Plane (Unicode) <https://en.wikipedia.org/wiki/Plane_(Unicode)>`.
+
+        Examples:
+            Below is a snippet of code that demonstrates how to use this function to convert
+            entities from Unicode to UTF-16 space. The `unicode_entities` are calculated in
```
Comment (Bibo-Joshi) [inline telegram/_messageentity.py:161]:
```suggestion
            entities from Unicode to UTF-16 space. The ``unicode_entities`` are calculated in
```

File: telegram/_messageentity.py [hunk 10]
```diff
@@ -140,6 +142,84 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+    ) -> Sequence["MessageEntity"]:
+        """Utility functionality for converting the offset and length of entities from
+        Unicode (`str`) to UTF-16 (`utf-16-le` encoded `bytes`).
+
+        Tip:
+            Only the offsets and lengths calulated in UTF-16 is acceptable by the Telegram Bot API.
+            If they are calculated using the Unicode string (`str` object), errors may occur
+            when the text contains characters that are not in the Basic Multilingual Plane (BMP).
+            For more information, see `Unicode <https://en.wikipedia.org/wiki/Unicode>` and
+            `Plane (Unicode) <https://en.wikipedia.org/wiki/Plane_(Unicode)>`.
```
Comment (Bibo-Joshi) [inline telegram/_messageentity.py:157]:
```suggestion
            For more information, see `Unicode <https://en.wikipedia.org/wiki/Unicode>`_ and
            `Plane (Unicode) <https://en.wikipedia.org/wiki/Plane_(Unicode)>`_.
```

File: telegram/_messageentity.py [hunk 11]
```diff
@@ -140,6 +142,84 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
 
         return super().de_json(data=data, bot=bot)
 
+    @staticmethod
+    def adjust_message_entities_to_utf_16(
+        text: str, entities: Sequence["MessageEntity"]
+    ) -> Sequence["MessageEntity"]:
+        """Utility functionality for converting the offset and length of entities from
+        Unicode (`str`) to UTF-16 (`utf-16-le` encoded `bytes`).
```
Comment (Bibo-Joshi) [inline telegram/_messageentity.py:150]:
```suggestion
        Unicode (:obj:`str`) to UTF-16 (``utf-16-le`` encoded :obj:`bytes`).
```

File: telegram/_messageentity.py [hunk 12]
```diff
@@ -140,6 +142,84 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+            text (:obj:`str`): The text that the entities belong to
+            entities (List[:class:`telegram.MessageEntity`]): Sequence of entities
+                with offset and length calculated in Unicode
+
+        Returns:
+            Sequence[:class:`telegram.MessageEntity`]: Sequence of entities
+            with offset and length calculated in UTF-16 encoding
+        """
+        # get sorted offsets
+        offsets_list = sorted(
```
Comment (Bibo-Joshi) [inline telegram/_messageentity.py:194]: I still find the usage of "offset" misleading both here and below. Please use change to "position"

File: telegram/_messageentity.py [hunk 13]
```diff
@@ -140,6 +142,84 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+            itertools.chain(*((x.offset, x.offset + x.length) for x in entities))
+        )
+        accumulated_length = 0
+        according_utf16_lengths: List[int] = []
```
Comment (Bibo-Joshi) [inline telegram/_messageentity.py:198]: I see no need for this helper variable. You can directly write the caculated postiions into "~offset~position_translation" as explained in https://github.com/python-telegram-bot/python-telegram-bot/pull/4323#discussion_r1655410569

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Thanks for the updates! Left some more comments, mostly about documentation :)
Please don't force-push additional changes, as that makes it hard to see what which parts I've already reviewed.

File: tests/test_messageentity.py [hunk 14]
```diff
@@ -81,6 +83,27 @@ def test_enum_init(self):
         entity = MessageEntity(type="url", offset=0, length=1)
         assert entity.type is MessageEntityType.URL
 
+    def test_fix_utf16(self):
+        import random
```
Comment (harshil21) [inline tests/test_messageentity.py:87]: please import at that the top of the module

File: telegram/_messageentity.py [hunk 15]
```diff
@@ -140,6 +142,84 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+        """Utility functionality for converting the offset and length of entities from
+        Unicode (`str`) to UTF-16 (`utf-16-le` encoded `bytes`).
+
+        Tip:
+            Only the offsets and lengths calulated in UTF-16 is acceptable by the Telegram Bot API.
+            If they are calculated using the Unicode string (`str` object), errors may occur
+            when the text contains characters that are not in the Basic Multilingual Plane (BMP).
+            For more information, see `Unicode <https://en.wikipedia.org/wiki/Unicode>` and
+            `Plane (Unicode) <https://en.wikipedia.org/wiki/Plane_(Unicode)>`.
+
```
Comment (harshil21) [inline telegram/_messageentity.py:158]:
```suggestion

        .. versionadded:: NEXT.VERSION
```
missing versionadded (please check my indentation)

**Review by Bibo-Joshi — APPROVED** [review]
Thanks for the additional updates! LGTM now :) If @harshil21 approves as well, we can merge.

File: telegram/_messageentity.py [hunk 16]
```diff
@@ -140,6 +142,81 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
...
+                    text=text,
+                    entities=utf_16_entities,
+                )
+                # utf_16_entities[0]: offset=3, length=4
+                # utf_16_entities[1]: offset=11, length=6
+                # utf_16_entities[2]: offset=30, length=6
+
+        Args:
+            text (:obj:`str`): The text that the entities belong to
+            entities (List[:class:`telegram.MessageEntity`]): Sequence of entities
```
Comment (harshil21) [inline telegram/_messageentity.py:188]:
```suggestion
            entities (Sequence[:class:`telegram.MessageEntity`]): Sequence of entities
```

**Review by harshil21 — CHANGES_REQUESTED** [review]
one last doc fix

**Comment by Bibo-Joshi** [comment]
Thank you very much for the nice contribution!

**Pull request merged**